### PR TITLE
(feat): `calculate_qc_metrics` with `dask`

### LINKED ISSUE #2937: `calculate_qc_metrics` for `dask`
### What kind of feature would you like to request?

Additional function parameters / changed functionality / changed defaults?

### Please describe your wishes

It would appear that the main blocker for `calculate_qc_metrics` being dask-ified is the fact that the calculation of top proportions is complicated and not easily dask-able.  So at the minimum, we can skip that and then do the rest of the computations which should be trivial.

### PR DESCRIPTION
**Pull request by ilan-gold**
- [x] Closes #2937
- [x] Tests included or not required because:

- [ ] Release notes not necessary because:

### CODE REVIEW

File: src/scanpy/preprocessing/_qc.py
Comment (ilan-gold) [inline src/scanpy/preprocessing/_qc.py]: This is the current blocker.   Without this commented, I just get a "quiet" failure.  Asked on zulip: https://scverse.zulipchat.com/#narrow/channel/456188-Dask/topic/Numba.20in.20Dask/near/478248965

**Comment by ilan-gold** [comment]
Apparently the problem I saw here with `numba` within dask is a bit more widespread: #3317

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]: maybe we should make and use a `getnnz` helper like `.._utils.axis_sum`?

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]: ditto

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]:
```suggestion
        adata.var[var_metrics.columns] = var_metrics
        return None
    return var_metrics
```

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]: please delete the docstring here, this is just a helper function to dispatch to, people should go through `top_segment_proportions`, right?

File: src/testing/scanpy/_helpers/__init__.py
Comment (flying-sheep) [inline src/testing/scanpy/_helpers/__init__.py]:
```suggestion
@contextmanager
def maybe_dask_process_context():
    """
    Running numba with dask's threaded scheduler causes crashes,
    so we need to switch to single-threaded (or processes, which is slower)
    scheduler for tests that use numba.
    """
    if not find_spec("dask"):
        yield
        return

    import dask.config

    prev_scheduler = dask.config.get("scheduler", "threads")
    dask.config.set(scheduler="single-threaded")
    try:
        yield
    finally:
        dask.config.set(scheduler=prev_scheduler)
```

**Review by flying-sheep — COMMENTED** [review]
looks great!

The diff for `test_qc_metrics` is a bit hard to follow. Am I right that you simply
1. pulled out `test_qc_metrics_no_log1p` and `test_qc_metrics_idempotent`
2. parametrized it with all array types
3. added `with maybe_dask_process_context():`

Thanks for pulling the tests apart!

**Comment by ilan-gold** [comment]
> Am I right that you simply

You are right!

File: src/scanpy/preprocessing/_qc.py
Comment (ilan-gold) [reply src/scanpy/preprocessing/_qc.py]: Ah great idea!

**Comment by flying-sheep** [comment]
great! then apart from my suggestions and [the crash problem](https://scverse.zulipchat.com/#narrow/channel/456188-Dask/topic/Numba.20in.20Dask/near/478897406), this looks ready.

File: src/scanpy/preprocessing/_qc.py
Comment (ilan-gold) [reply src/scanpy/preprocessing/_qc.py]: why should we delete the docstring, seems better than not having one?

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]: This should probably be collapsed into something like:

```py
obs_metrics[f"n_{var_type}_by_{expr_type}"] = materialize_as_ndarray(axis_nnz(X, axis=1))
```

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py]: same here, the whole elif tree can probably become something like

```py
var_metrics["n_cells_by_{expr_type}"], var_metrics["mean_{expr_type}"] = (
    materialize_as_ndarray(axis_nnz(X, axis=0), _get_mean_var(X, axis=0)[0])
)
```

File: src/scanpy/_utils/__init__.py
Comment (flying-sheep) [inline src/scanpy/_utils/__init__.py:717]: maybe `X: ArrayLike`? With the idea that this is the fallback branch that’s able to handle everything that numpy can convert into an array?

Comment (flying-sheep) [inline src/scanpy/_utils/__init__.py]: maybe `X: ArrayLike` (and still `-> np.ndarray`)? With the idea being that this is the fallback branch that’s able to handle everything that numpy can convert into an array?

File: src/scanpy/_utils/__init__.py
Comment (flying-sheep) [inline src/scanpy/_utils/__init__.py:728]: strange, codecov says this line isn’t covered.

I assume it’s just lying, but please check if that message has vanished before you merge.

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [inline src/scanpy/preprocessing/_qc.py:197]: I’m just realizing that none of these are `f`-literals. `expr_type` won’t be interpolated into this!

mayyybe we should fix that.

File: src/scanpy/preprocessing/_qc.py
Comment (flying-sheep) [reply src/scanpy/preprocessing/_qc.py]: it’s duplicated, so when someone modifies it, there’s a >0 chance that they only modify one version and not both, which is worse than having to search for the docstring for a few seconds.

if you want, you can rename these internal helper functions to `_leading_underscore` names, then it’s even more likely that future us will find the entrypoint `top_proportions` and its docstring.

**Review by flying-sheep — APPROVED** [review]
beautiful. we should decide how to deal with the missing `f`-literals, otherwise LGTM

File: src/scanpy/preprocessing/_qc.py
Comment (ilan-gold) [reply src/scanpy/preprocessing/_qc.py:197]: i think they are! `col.format` is called!

Comment (flying-sheep) [reply src/scanpy/preprocessing/_qc.py:197]: ah! bizarre! but sure, let’s do that in another PR then

File: src/scanpy/_utils/__init__.py
Comment (ilan-gold) [reply src/scanpy/_utils/__init__.py:728]: Screenshot 2024-10-30 at 09 49 45
gone!

File: src/scanpy/preprocessing/_qc.py
Comment (ilan-gold) [reply src/scanpy/preprocessing/_qc.py]: great point, refactored for singledispatch

**Comment by flying-sheep** [comment]
OK, https://github.com/scverse/scanpy/pull/3335 is ready.

We should manually check if it can get rid of `maybe_dask_process_context`

/edit: it’s not ready

**Pull request merged**
